flang/lib/Lower/Bridge.cpp

```diff
@@ -525,33 +544,39 @@ class FirConverter : public Fortran::lower::AbstractConverter {
   /// Generate the address of loop variable \p sym.
   /// If \p sym is not mapped yet, allocate local storage for it.
   mlir::Value genLoopVariableAddress(mlir::Location loc,
-                                     const Fortran::semantics::Symbol &sym) {
-    if (lookupSymbol(sym)) {
-      Fortran::lower::StatementContext stmtCtx;
-      return fir::getBase(
-          genExprAddr(Fortran::evaluate::AsGenericExpr(sym).value(), stmtCtx));
+                                     const Fortran::semantics::Symbol &sym,
+                                     bool isUnordered) {
+    if (isUnordered || sym.has<Fortran::semantics::HostAssocDetails>() ||
```

Comment:
Why is a shallow lookup required for the unordered case ?

Reply:
It catches the DO CONCURRENT case, where we were adding a new symbol to the map for the variable in the concurrent header block.

Reply:
Thanks, makes sense now.